src/librustdoc/visit_ast.rs

```diff
@@ -61,20 +61,60 @@ impl<'a, 'tcx> RustdocVisitor<'a, 'tcx> {
         );
-        // Attach the crate's exported macros to the top-level module:
-        module.macros.extend(krate.exported_macros.iter().map(|def| (def, None)));
-        module.is_crate = true;
-
+        top_level_module.is_crate = true;
+        // Attach the crate's exported macros to the top-level module.
+        // In the case of macros 2.0 (`pub macro`), and for built-in `derive`s or attributes as
+        // well (_e.g._, `Copy`), these are wrongly bundled in there too, so we need to fix that by
+        // moving them back to their correct locations.
+        'exported_macros: for def in krate.exported_macros {
+            // The `def` of a macro in `exported_macros` should correspond to either:
+            //  - a `#[macro_export] macro_rules!` macro,
+            //  - a built-in `derive` (or attribute) macro such as the ones in `::core`,
+            //  - a `pub macro`.
+            // Only the last two need to be fixed, thus:
+            if def.ast.macro_rules {
+                top_level_module.macros.push((def, None));
+                continue 'exported_macros;
+            }
+            let tcx = self.cx.tcx;
+            // Note: this is not the same as `.parent_module()`. Indeed, the latter looks
+            // for the closest module _ancestor_, which is not necessarily a direct parent
+            // (since a direct parent isn't necessarily a module, c.f. #77828).
+            let macro_parent_def_id = {
+                use rustc_middle::ty::DefIdTree;
+                tcx.parent(tcx.hir().local_def_id(def.hir_id).to_def_id()).unwrap()
+            };
+            let macro_parent_path = tcx.def_path(macro_parent_def_id);
+            // HACK: rustdoc has no way to lookup `doctree::Module`s by their HirId. Instead,
+            // lookup the module by its name, by looking at each path segment one at a time.
+            let mut cur_mod = &mut top_level_module;
+            for path_segment in macro_parent_path.data {
+                // Path segments may refer to a module (in which case they belong to the type
+                // namespace), which is _necessary_ for the macro to be accessible outside it
+                // (no "associated macros" as of yet). Else we bail with an outer `continue`.
+                let path_segment_ty_ns = match path_segment.data {
+                    rustc_hir::definitions::DefPathData::TypeNs(symbol) => symbol,
+                    _ => continue 'exported_macros,
+                };
+                // Descend into the child module that matches this path segment (if any).
+                match cur_mod.mods.iter_mut().find(|child| child.name == Some(path_segment_ty_ns)) {
+                    Some(child_mod) => cur_mod = &mut *child_mod,
+                    None => continue 'exported_macros,
+                }
+            }
+            let cur_mod_def_id = tcx.hir().local_def_id(cur_mod.id).to_def_id();
+            assert_eq!(cur_mod_def_id, macro_parent_def_id);
```

Comment:
Now that you added the comment, I think this isn't true in general and there will be times the assertion fails, because `cur_mod_def_id` is *always* a module and `macro_parent` may not be. I think looking for the nearest module is the correct behavior.

Reply:
Technically that should be caught by the loop logic: if one of the ancestors is not a module, we will be hitting one of the two `continue 'exported macros;` bail points, and thus never reach that part of the code. That being said, I am open to suggestions to make these things clearer (_e.g._, we could `.try_fold()` and `Err` rather than `for + continue` 🤷)

Reply:
Ok, but wouldn't it be simpler to only look up the parent module in the first place? Then you wouldn't have to skip the segments you don't care about, because you care about all of them.

Reply:
We could add a "fail faster" branch where we'd compare the immediate parent and the parent module, and bail if they're not equal; but even then we'd still need to check that that module in and of itself is reachable / nameable, so most of the logic would still be here. I think the micro optimization is not worth the extra branch; but I keep thinking that a `try_fold()` for the inner loop may be interesting, since it avoids the need to `continue` the `'outer` loop:

<details><summary>Example (assuming postfix <code>match</code> for the ergonomics)</summary>

```rust
for def in krate.exported_macros {
    let put_macro_in = |module| module.macros.push((def, None));
    // The `def` of a macro in `exported_macros` should correspond to either:
    //  - a `#[macro_export] macro_rules!` macro,
    //  - a built-in `derive` (or attribute) macro such as the ones in `::core`,
    //  - a `pub macro`.
    // Only the last two need to be fixed, thus:
    if def.ast.macro_rules {
        put_macro_in(&mut top_level_module);
        continue;
    }
    let tcx = self.cx.tcx;
    // Note: this is not the same as `.parent_module()`. Indeed, the latter looks
    // for the closest module _ancestor_, which is not necessarily a direct parent
    // (since a direct parent isn't necessarily a module, c.f. #77828).
    let macro_parent_def_id = {
        use rustc_middle::ty::DefIdTree;
        tcx.parent(tcx.hir().local_def_id(def.hir_id).to_def_id()).unwrap()
    };
    let macro_parent_path = tcx.def_path(macro_parent_def_id);
    // HACK: rustdoc has no way to lookup `doctree::Module`s by their HirId. Instead,
    // lookup the module by its name, by looking at each path segment one at a time.
    macro_parent_path.data.try_fold(&mut top_level_module, |(cur_mod, path_segment)| {
        // Path segments may refer to a module (in which case they belong to the type
        // namespace), which is _necessary_ for the macro to be accessible outside it
        // (no "associated macros" as of yet). Else we bail.
        let path_segment_ty_ns = match path_segment.data {
            rustc_hir::definitions::DefPathData::TypeNs(symbol) => symbol,
            _ => return Err(()),
        };
        // Descend into the child module that matches this path segment, if any.
        cur_mod.mods.iter_mut().find(|child| child.name == Some(path_segment_ty_ns)).ok_or(())
    }).match {
        Ok(module) => {
            let module_def_id =
                tcx.hir().local_def_id(module.id).to_def_id()
            ;
            assert_eq!(module_def_id, macro_parent_def_id);
            put_macro_in(&mut module);
        },
        Err(()) => continue,
    }
}
```

</details>

Reply:
Hmm, ok. I'm fine with either the current code or the try_fold.